sympy/functions/special/holonomic.py

```diff
@@ -0,0 +1,429 @@
+"""Holonomic Functions and Differential Operators"""
+
+from __future__ import print_function, division
+
+from sympy import symbols, Symbol, diff, S, Mul
+from sympy.polys.polytools import lcm, gcd
+from sympy.core.expr import Expr
+from sympy.printing import sstr
+from sympy.matrices import Matrix
+from sympy.core.compatibility import range
+from sympy.polys.polytools import DMP
+
+
+def DiffOperatorAlgebra(base_ring, generator):
+    """ A function to create a Differential Operator Algebra.
+    The first arguments need to be the base polynomial ring for the algebra
+    and the second argument must be a generator.
+    Returns the class representing the operator algebra and
+    the standard operator for differentiation i.e. the `Dx` operator.
+
+    Examples
+    =======
+    from sympy.polys.domains import ZZ
+    R, Dx = DiffOperatorAlgebra
+    """
+
+    ring = DifferentialOperatorAlgebra(base_ring, generator)
+    return (ring, ring.diff_operator)
+
+
+class DifferentialOperatorAlgebra:
+    """ The class representing Differential Operator Algebra.
+    Defined by the base polynomial ring and the generator.
+
+    The attribute diff_operator is the operator `Dx` which when acted
+    on a function does differentiation.
+    """
+
+    def __init__(self, base_ring, generator):
+
+        self.base_ring = base_ring
+        self.diff_operator = DifferentialOperator([0, 1], self, generator)
```

Comment:
0 and 1 could also be ``base_ring.zero`` and ``base_ring.one`` which have the correct type.